src/osp/Universe.cpp

```diff
@@ -49,27 +53,61 @@ void Universe::sat_remove(Satellite sat)
     m_registry.destroy(sat);
 }
 
-Vector3s Universe::sat_calc_pos(Satellite referenceFrame, Satellite target) const
+Vector3g Universe::sat_calc_pos(Satellite referenceFrame, Satellite target) const
 {
-    auto const view = m_registry.view<const UCompTransformTraj>();
+    auto const viewInCoord = m_registry.view<const UCompInCoordspace>();
+    auto const viewCoordIndex = m_registry.view<const UCompCoordspaceIndex>();
 
     // TODO: maybe do some checks to make sure they have the components
-    auto const &framePosTraj = view.get<const UCompTransformTraj>(referenceFrame);
-    auto const &targetPosTraj = view.get<const UCompTransformTraj>(target);
+    auto const &frameInCoord
+            = viewInCoord.get<const UCompInCoordspace>(referenceFrame);
+    auto const &targetInCoord
+            = viewInCoord.get<const UCompInCoordspace>(target);
 
-    if (framePosTraj.m_parent == targetPosTraj.m_parent)
+    auto const &frameCoordIndex
+            = viewCoordIndex.get<const UCompCoordspaceIndex>(referenceFrame);
+    auto const &targetCoordIndex
+            = viewCoordIndex.get<const UCompCoordspaceIndex>(target);
+
+    if (frameInCoord.m_coordSpace == targetInCoord.m_coordSpace)
     {
-        // Same parent, easy calculation
-        return targetPosTraj.m_position - framePosTraj.m_position;
+        // Both Satellites are in the same coordinate system
+
+        std::optional<CoordinateSpace> const &rSpace
+                = m_coordSpaces.at(frameInCoord.m_coordSpace);
+
+        ViewCComp_t<CCompX> const& viewX = rSpace->ccomp_view<CCompX>();
+        ViewCComp_t<CCompY> const& viewY = rSpace->ccomp_view<CCompY>();
+        ViewCComp_t<CCompZ> const& viewZ = rSpace->ccomp_view<CCompZ>();
+
+        Vector3g const pos{
+            viewX[targetCoordIndex.m_myIndex] - viewX[frameCoordIndex.m_myIndex],
+            viewY[targetCoordIndex.m_myIndex] - viewY[frameCoordIndex.m_myIndex],
+            viewZ[targetCoordIndex.m_myIndex] - viewZ[frameCoordIndex.m_myIndex],
+        };
+
+        return pos;
     }
-    // TODO: calculation for different parents
+
+    // TODO: calculation for positions across different coordinate spaces
```

Comment:
Does this include objects that are transitioning from one coordinate space (which I think i once referred to as a sphere of influence a long time ago) to another, such as leaving orbit of a planet / moon, and thus transitioning to the orbit of the parent body (star / planet) ?

Reply:
Satellites in mid-transition (I personally call it a handoff) would probably be in an invalid state.

A Trajectory function would be responsible for determining when it's time to give a Satellite to a new coordinate space. It has to calculate the position of the Satellite relative to the new coordinate space (maybe a different function would be more efficient), then throw it into the new coordinate space's mailbox with add(...).

Reply:
Do you know where you'd like that to happen in the per-frame order of operations?

Reply:
This is going to be put in the upcoming universe updates. I guess there are two main steps:
1. Trajectory function moves the Satellites, and also determine which satellites need to be handed off to other coordinate systems using queues
2. Exchange (like CoordspaceCartesianSimple::update_exchange) would add and remove the satellites coordinates from their buffers